File: src/main/java/net/draconia/askaround/domain/City.java
package net.draconia.askaround.domain;

import java.io.Serializable;

public class City implements Serializable
{
	private static final long serialVersionUID = 6298054579376482225L;
	
	private Integer miId;
	private State mObjState;
	private String msName;
	
	public City()
	{ }
	
	public City(final String sName, final State objState)
	{
		setName(sName);
		setState(objState);
	}
	
	public City(final Integer iId, final String sName, final State objState)
	{
		setId(iId);
		setName(sName);
		setState(objState);
	}
	
	public int getId()
	{
		if(miId == null)
			miId = 0;
		
		return(miId);
	}
	
	public String getName()
	{
		if(msName == null)
			msName = "";
		
		return(msName);
	}
	
	public State getState()
	{
		if(mObjState == null)
			mObjState = new State();
		
		return(mObjState);
	}
	
	public void setId(final Integer iId)
	{
		if(iId == null)
			miId = 0;
		else
			miId = iId;
	}
	
	public void setName(final String sName)
	{
		if(sName == null)
			msName = "";
		else
			msName = sName;
	}
	
	public void setState(final State objState)
	{
		if(objState == null)
			mObjState = new State();
		else
			mObjState = objState;
	}
}

File: src/main/webapp/scripts/models/city.js
function City()
{
	this.mbChanged						=	undefined;
	this.miId							=	undefined;
	this.msName							=	undefined;
	this.mArrObjObservers				=	[];
	this.mObjState						=	undefined;
}

City.prototype.addObserver				=	function(objObserver)
{
	if(objObserver != undefined)
		{
		if(this.mArrObjObservers == undefined)
			this.mArrObjObservers 		=	[];
		
		this.mArrObjObservers.push(objObserver);
		}
	
	return(this);
};

City.copyFromJSON					=	function(objJSON, objCity)
{
	if(objCity == undefined)
		objCity						=	new City();
	
	for(prop in objJSON)
		switch(prop)
			{
			case "state":
				objCity.setState(State.copyFromJSON(eval("objJSON." + prop), objCity.getState()));
				break;
			default:
				eval("objCity.set" + prop.substring(0, 1).toUpperCase() + prop.substring(1) + "(objJSON." + prop + ");");
				break;
			}
	
	return(objCity);
};

City.prototype.getId					=	function()
{
	if(this.miId == undefined)
		this.miId						=	0;
	
	return(this.miId);
};

City.prototype.getName					=	function()
{
	if(this.msName == undefined)
		this.msName						=	"";
	
	return(this.msName);
};

City.prototype.getState					=	function()
{
	if(this.mObjState == undefined)
		this.mObjState					=	new State();
	
	return(this.mObjState);
};

City.prototype.hasChanged				=	function()
{
	if(this.mbChanged == undefined)
		this.mbChanged 					=	false;
	
	return(this.mbChanged);
};

City.prototype.notifyObservers			=	function(objArgument)
{
	this.mbChanged = true;
	
	for(var iLength = this.mArrObjObservers.length, iLoop = 0; iLoop < iLength; iLoop++)
		this.mArrObjObservers[iLoop].call(this, objArgument);
	
	this.mbChanged = false;
};

City.prototype.packageToJSON			=	function()
{
	return(	{	id:		this.getId()
			,	name:	this.getName()
			,	state:	this.getState().packageToJSON()
			});
};

City.prototype.setId					=	function(iId)
{
	if(iId == undefined)
		this.miId						=	0;
	else
		this.miId						=	iId;
	
	this.notifyObservers();
	
	return(this);
};

City.prototype.setName					=	function(sName)
{
	if(sName == undefined)
		this.msName						=	"";
	else
		this.msName						=	sName;
	
	this.notifyObservers();
	
	return(this);
};

City.prototype.setState					=	function(objState)
{
	if(objState == undefined)
		this.mObjState					=	new State();
	else
		this.mObjState					=	objState;
	
	this.notifyObservers();
	
	return(this);
};